telemetry/telemetry.go

```diff
@@ -29,48 +31,80 @@ func NewTelemetry(cfg *Config) *Telemetry {
 	}
 	t := &Telemetry{
 		Path:    "/metrics", // TODO hard-coded?
-		lock:    sync.RWMutex{},
 		Sensors: []*Sensor{},
 	}
 	t.addr = cfg.addr
-	t.mux = http.NewServeMux()
-	t.mux.Handle(t.Path, prometheus.Handler())
+	router := http.NewServeMux()
+	router.Handle(t.Path, prometheus.Handler())
+	t.Handler = router
+
 	for _, sensorCfg := range cfg.SensorConfigs {
 		sensor := NewSensor(sensorCfg)
 		t.Sensors = append(t.Sensors, sensor)
 	}
+	t.Rx = make(chan events.Event, 10)
 	return t
 }
 
-var listener net.Listener
+// Run executes the event loop for the telemetry server
+func (t *Telemetry) Run(bus *events.EventBus) {
+	t.Subscribe(bus, true)
+	t.Bus = bus
+	t.Start()
 
-// Serve starts serving the telemetry service
-func (t *Telemetry) Serve() {
-	t.lock.Lock()
-	defer t.lock.Unlock()
+	go func() {
+		defer t.Stop()
+		for {
+			event := <-t.Rx
+			switch event {
+			case
+				events.QuitByClose,
+				events.GlobalShutdown:
+				return
+			}
+		}
+	}()
+}
 
-	// No-op if we've created the server previously.
-	// TODO: golang's native implementation of http.Server.Server() cannot
-	// support graceful reload. We need to select an alternate implementation
-	// but in the meantime we need to back-out the change to reloading
-	// ref https://github.com/joyent/containerpilot/pull/165
-	if listener != nil {
-		return
-	}
-	ln, err := net.Listen(t.addr.Network(), t.addr.String())
-	if err != nil {
-		log.Fatalf("Error serving telemetry on %s: %v", t.addr.String(), err)
-	}
-	listener = ln
-	t.listening = true
+// Start starts serving the telemetry service
+func (t *Telemetry) Start() {
+	ln := t.listenWithRetry()
```

Comment:
Yah. If we had 3 servers I'd say "let's extract all this out into a server+eventhandler abstraction" but it doesn't seem like it's worth the effort or indirection with just the 2.